Kernel/FileSystem/FATFileSystem.cpp

```diff
@@ -0,0 +1,358 @@
+            lfn_entries.clear();
+        }
+    }
+
+    return EINVAL;
+}
+
+ErrorOr<void> FATInode::compute_filename(Vector<FATLongFileNameEntry> lfn_entries)
+{
+    VERIFY(!m_filename);
+
+    if (lfn_entries.is_empty()) {
+        StringBuilder filename;
+        filename.append(byte_terminated_string(reinterpret_cast<u8*>(&m_entry.filename), 8, ' '));
+        if (m_entry.extension[0] != ' ') {
+            filename.append('.');
+            filename.append(byte_terminated_string(reinterpret_cast<u8*>(&m_entry.extension), 3, ' '));
+        }
+        m_filename = TRY(KString::try_create(filename.string_view()));
+    } else {
+        StringBuilder filename;
+        for (auto& lfn_entry : lfn_entries) {
+            filename.append(lfn_entry.characters1[0]);
+            filename.append(lfn_entry.characters1[1]);
+            filename.append(lfn_entry.characters1[2]);
+            filename.append(lfn_entry.characters1[3]);
+            filename.append(lfn_entry.characters1[4]);
+            filename.append(lfn_entry.characters2[0]);
+            filename.append(lfn_entry.characters2[1]);
+            filename.append(lfn_entry.characters2[2]);
+            filename.append(lfn_entry.characters2[3]);
+            filename.append(lfn_entry.characters2[4]);
+            filename.append(lfn_entry.characters2[5]);
+            filename.append(lfn_entry.characters3[0]);
+            filename.append(lfn_entry.characters3[1]);
+        }
+
+        m_filename = TRY(KString::try_create(byte_terminated_string(filename.string_view().bytes().data(), filename.length(), 0xFF)));
+    }
+
+    dbgln_if(FAT_DEBUG, "FATFS: Computed filename {} for inode {}", m_filename, index());
+    return {};
+}
+
+time_t FATInode::fat_date_time(u16 date, u16 time) const
+{
+    time_t year = (date >> 9) + (1980 - 1970);
+    time_t month = (date >> 5) & 0b1111;
+    time_t day = date & 0b11111;
+
+    time_t hour = time >> 11;
+    time_t minute = (time >> 5) & 0b1111111;
+    time_t second = (time & 0b11111) << 1;
+
+    return second + minute * 60 + hour * 3600
+        + day + (24 * 3600) + month * (30 * 24 * 3600) + year * (365 * 24 * 3600);
+}
+
+// TODO: This might be worth putting as a part of the StringView class
+StringView FATInode::byte_terminated_string(u8 const* initial_string, size_t max_length, u8 byte) const
+{
+    StringView string = StringView(initial_string, max_length);
+
+    while (string.length() > 0) {
+        if (string[string.length() - 1] != byte)
+            break;
+        string = string.substring_view(0, string.length() - 1);
+    }
+
+    return string;
+}
+
+ErrorOr<size_t> FATInode::read_bytes_locked(off_t offset, size_t size, UserOrKernelBuffer& buffer, OpenFileDescription*) const
+{
+    dbgln_if(FAT_DEBUG, "FATFS: Reading inode {}: ize: {} offset: {}", identifier().index(), size, offset);
+
+    // FIXME: Read only the needed blocks instead of the whole file
+    auto blocks = TRY(const_cast<FATInode&>(*this).read_block_list());
```

Comment:
That looks very weird. Maybe move `generate_block_list` somewhere else and make `read_block_list` `const` as well.

Reply:
There isn't an easy way to make `read_block_list` `const`. It uses some read functions that are not marked `const` (maybe they should be?) and also I think we should only compute the block list if we need in place like here, because let's say you want to list a directory with a very large amount of files, it would be pointless to run this function for every one of the files because we only about their file name.